### ISSUE TITLE
[ao/pulse] --audio-device doesn't work

### ISSUE THREAD

**Issue by GUiHKX** [mpv-player/mpv#5482]
### mpv version and platform

```
mpv 0.28.0-189-g7f3c7100d5 Copyright © 2000-2018 mpv/MPlayer/mplayer2 projects
 built on UNKNOWN
ffmpeg library versions:
   libavutil       56.7.100
   libavcodec      58.9.100
   libavformat     58.7.100
   libswscale      5.0.101
   libavfilter     7.11.101
   libswresample   3.0.101
ffmpeg version: N-89945-g108958e43d
```

### Reproduction steps

I have two output devices: Headphones and HDMI (NVIDIA). I want to use the HDMI output, but even if I specifically set it using `--audio-device`, mpv still uses my headphones:

```
mpv --audio-device=help | grep pulse
  'pulse/alsa_output.pci-0000_01_00.1.hdmi-stereo-extra1' (GK106 HDMI Audio Controller Digital Stereo (HDMI 2))
  'pulse/alsa_output.pci-0000_00_1b.0.analog-stereo' (Built-in Audio Analog Stereo)
```
Then:
```
mpv --no-config --audio-device=pulse/alsa_output.pci-0000_01_00.1.hdmi-stereo-extra1 song.flac --log-file=ao.txt
Playing: song.flac
 (+) Video --vid=1 [P] (mjpeg 1200x1200)
 (+) Audio --aid=1 (flac 2ch 44100Hz)
File tags:
 Artist: Breaking Benjamin
 Album: Saturate
 Album_Artist: Breaking Benjamin
 Genre: Rock
 Performer: Breaking Benjamin
 Title: Forever
 Track: 13
Displaying attached picture. Use --no-audio-display to prevent this.
VO: [gpu] 1200x1200 yuv444p
AO: [pulse] 44100Hz stereo 2ch s16
AV: 00:00:01 / 00:03:54 (0%)


Exiting... (Quit)
```

If I open `pavucontrol` and manually change the audio device used by mpv, it works.

### Expected behavior
mpv correctly uses HDMI as my audio device.

### Actual behavior
mpv outputs audio to headphones instead.

### Log file
[ao.txt](https://github.com/mpv-player/mpv/files/1690349/ao.txt)

**Comment by thebombzen** [comment]
I cannot reproduce this in git master using my Nvidia HDMI output as you did above. (Using `--audio-device=pulse/foo_hdmi_bar`.)

Check your PulseAudio configuration. Does the "Set as fallback" option affect anything? This checkbox:
![Pavucontrol Dialog](https://0x0.st/sb7b.png)

**Comment by GUiHKX** [comment]
Unfortunately It has no effect at all. When you have some free time, see if you can watch this video I made while reproducing the issue; if you spot something misconfigured in pavucontrol, let me know (I'm kinda of a noob):

https://www.youtube.com/watch?v=mFy2Ok_BN2w&feature=youtu.be

It's weird because that's also happening in mpv 0.27.0, and I've **never** had that issue before. I've recently switched my desktop environment (from XFCE to KDE), but when I used mpv on XFCE, switching between output devices worked just fine (I used [this script](https://gist.github.com/bitingsock/ad58ee5da560ecb922fa4a867ac0ecfd) to make it easier). Obviously, I actually have no idea if this is related to KDE, but I'm almost certain it started to happening after the switch.

I'm willing to reset PulseAudio (if that's a thing), but even if that works, wouldn't that bug (or whatever that is) still be not fixed?

**Comment by thebombzen** [comment]
I'm using MATE, another GTK-based desktop and my behavior is similar to yours in XFCE. It's possible this is a bug in how PulseAudio behaves when KDE is around. 

My best guess is that mpv is correctly setting the output audio device and somewhere in Pulse's web of crap, it's being routed to your headphones. Is this reproduceable with ALSA audio out?

**Comment by GUiHKX** [comment]
I've tried all alsa outputs since I was not sure which was the correct one:

```
$ mpv --audio-device=help | grep alsa/
  'alsa/sysdefault:CARD=PCH' (HDA Intel PCH, ALC662 rev1 Analog/Default Audio Device)
  'alsa/front:CARD=PCH,DEV=0' (HDA Intel PCH, ALC662 rev1 Analog/Front speakers)
  'alsa/surround21:CARD=PCH,DEV=0' (HDA Intel PCH, ALC662 rev1 Analog/2.1 Surround output to Front and Subwoofer speakers)
  'alsa/surround40:CARD=PCH,DEV=0' (HDA Intel PCH, ALC662 rev1 Analog/4.0 Surround output to Front and Rear speakers)
  'alsa/surround41:CARD=PCH,DEV=0' (HDA Intel PCH, ALC662 rev1 Analog/4.1 Surround output to Front, Rear and Subwoofer speakers)
  'alsa/surround50:CARD=PCH,DEV=0' (HDA Intel PCH, ALC662 rev1 Analog/5.0 Surround output to Front, Center and Rear speakers)
  'alsa/surround51:CARD=PCH,DEV=0' (HDA Intel PCH, ALC662 rev1 Analog/5.1 Surround output to Front, Center, Rear and Subwoofer speakers)
  'alsa/surround71:CARD=PCH,DEV=0' (HDA Intel PCH, ALC662 rev1 Analog/7.1 Surround output to Front, Center, Side, Rear and Woofer speakers)
  'alsa/hdmi:CARD=NVidia,DEV=0' (HDA NVidia, HDMI 0/HDMI Audio Output)
  'alsa/hdmi:CARD=NVidia,DEV=1' (HDA NVidia, HDMI 1/HDMI Audio Output)
  'alsa/hdmi:CARD=NVidia,DEV=2' (HDA NVidia, HDMI 2/HDMI Audio Output)
  'alsa/hdmi:CARD=NVidia,DEV=3' (HDA NVidia, HDMI 3/HDMI Audio Output)
```

None of them outputs to HDMI; some of them outputs to headphones; and the rest of them gives me this error:

```
[ao/alsa] Playback open error: Device or resource busy
[ao] Failed to initialize audio driver 'alsa'                                                                                             
[ao] This audio driver/device was forced with the --audio-device option.                                                                  
[ao] Try unsetting it.                                                                                                                    
Could not open/initialize audio device -> no sound.
```

**Comment by GUiHKX** [comment]
I've reinstalled XFCE; It didn't work, I was still getting the same behavior I have on KDE. However, I remembered I had installed this program called `pasystray` while I was on XFCE, because the built-in plug-in for volume control on XFCE is really limited IMO. So I installed pasystray-git and tried to reproduce it again, and **now changing audio output via command-line, script or pavucontrol WORKS**. 

I went and tried it on KDE (I left pasystray installed, still), but it didn't work on KDE. I don't know what pasystray does, I simply thought it was a panel plug-in to make audio output switching easier. 

Now that I know the issue was somehow caused by pasystray, I also know that simply uninstalling it wouldn't solve the problem because I had done that before, so what I did was move PulseAudio's current configuration files to another folder, then PulseAudio would generate new ones:

```
$ mv ~/.config/pulse ~/.config/pulse.old
$ pulseaudio --kill && pulseaudio --start
```

That's it. Switching between audio devices on mpv works properly now.

**TL;DR: Somehow the issue was caused by pasystray earlier when I had it installed on XFCE, uninstalling it and removing PulseAudio config. files fixed this issue. **

I'm sorry for wasting your guys' time!

**Comment by GUiHKX** [comment]
@kevmitch Thanks for the suggestion, but I guess it's not even worth trying that now. Sorry again :+1:

**Comment by kevmitch** [comment]
no worries, posted that before i saw you closed the issue.

**Comment by nanderer** [comment]
my problem is that when my system overloads von time to time (ram/cpu near 100%) pulseaudio crashes and then all running mpv's (i have a couple of things i usually started watching and paused, never rebooting unless i have to) 
```
ps aux | grep mpv |wc
     23    1378   16133
```
all of them connect themself to alsa, even if i am not playing them, and then when the overloaded condition is gone (usually using ssh and killing firefox or just going to bed when i cant access via ssh and waiting for oom killer during the night) 

then when pulseaudio restarts it is not able to use my card anymore cause all mpvs grapped, pulseaudio cant access alsa-card anymore, i substitute this then with alsa-sink/alsa-source but it would be much better if mpv would not take alsa if it was connected to pulse before unless in playing state, so pulse could restart and keep running in low latance mode.

what i do then after pulse runs again is enabling this line in my mpv.conf

```
#temporary while pulse is not working as normal
audio-device=alsa/sysdefault
```

so i dont have to cope with the latency using alsa-sink
